File: shell/plasma/package/contents/ui/logic/UpdateCheck.js
.pragma library

.import "FormatTime.js" as FormatTime
.import "I18n.js" as I18n

const STATUSES = ["up_to_date", "available", "failed", "rate_limited", "disabled"];
const MINIMUM_VERSION = [0, 6, 0];
const RETRY_ERRORS = ["org.freedesktop.DBus.Error.NoReply", "org.freedesktop.DBus.Error.Timeout"];
const NOT_SUPPORTED = "org.freedesktop.DBus.Error.NotSupported";
const SERVICE_UNKNOWN = "org.freedesktop.DBus.Error.ServiceUnknown";
const MAX_ATTEMPTS = 3;
const KEPT_RESULTS = ["failed", "rate_limited", "disabled"];

function isObject(value) {
    return value !== null && typeof value === "object" && !Array.isArray(value);
}

function text(value) {
    return typeof value === "string" && value.trim().length > 0 ? value.trim() : null;
}

function timestamp(value) {
    const ms = typeof value === "string" ? Date.parse(value) : NaN;
    return Number.isNaN(ms) ? null : new Date(ms);
}

function parseUpdateCheck(raw) {
    if (!isObject(raw))
        return null;
    return {
        checkedAt: timestamp(raw.checked_at)
    };
}

function result(status, checkedAt, version, until) {
    return {
        status,
        checkedAt,
        version,
        until
    };
}

function failed(checkedAt) {
    return result("failed", checkedAt, null, null);
}

function parseJson(json) {
    try {
        return JSON.parse(json);
    } catch (error) {
        return null;
    }
}

function parseResult(json, previousCheckedAt) {
    const raw = parseJson(json);
    if (!isObject(raw) || !STATUSES.includes(raw.status))
        return failed(previousCheckedAt);
    return result(raw.status, timestamp(raw.checked_at), text(raw.version), timestamp(raw.until));
}

function outcome(errorName, value, previousCheckedAt) {
    if (errorName === "")
        return parseResult(value, previousCheckedAt);
    if (errorName === NOT_SUPPORTED)
        return result("disabled", null, null, null);
    return failed(previousCheckedAt);
}

function retries(errorName, attempt) {
    return RETRY_ERRORS.includes(errorName) && attempt < MAX_ATTEMPTS;
}

function versionParts(version) {
    const match = /^v?(\d+)\.(\d+)\.(\d+)/.exec(version ?? "");
    return match === null ? null : [Number(match[1]), Number(match[2]), Number(match[3])];
}

function supported(appVersion) {
    const parts = versionParts(appVersion);
    if (parts === null)
        return false;
    const differing = parts.findIndex((part, index) => part !== MINIMUM_VERSION[index]);
    return differing === -1 || parts[differing] > MINIMUM_VERSION[differing];
}

function visible(checksOn, snapshot) {
    return checksOn && snapshot !== null && snapshot.updateCheck !== null && supported(snapshot.appVersion);
}

function fromState(snapshot) {
    const checkedAt = snapshot.updateCheck?.checkedAt ?? null;
    if (snapshot.update !== null)
        return result("available", checkedAt, snapshot.update.version, null);
    return result(checkedAt === null ? "unchecked" : "up_to_date", checkedAt, null, null);
}

function shown(snapshot, lastResult) {
    if (lastResult !== null && KEPT_RESULTS.includes(lastResult.status))
        return lastResult;
    return fromState(snapshot);
}

function checked(lang, checkedAt, now) {
    return I18n.tr(lang, "checked {ago}", {
        ago: FormatTime.agoText(lang, checkedAt, now)
    });
}

function rateLimitedLine(lang, until, now) {
    if (until === null || until <= now)
        return I18n.tr(lang, "GitHub limits update checks · try again later");
    return I18n.tr(lang, "GitHub limits update checks · try again in {duration}", {
        duration: FormatTime.duration(lang, until - now)
    });
}

function upToDateLine(lang, appVersion, checkedAt, now) {
    return I18n.tr(lang, "You're up to date · Headroom {version} · {checked}", {
        version: appVersion,
        checked: checked(lang, checkedAt, now)
    });
}

function availableLine(lang, checkedAt, now) {
    if (checkedAt === null)
        return I18n.tr(lang, "Update available");
    return I18n.tr(lang, "Update available · {checked}", {
        checked: checked(lang, checkedAt, now)
    });
}

function failedLine(lang, checkedAt, now) {
    if (checkedAt === null)
        return I18n.tr(lang, "Couldn't check for updates");
    return I18n.tr(lang, "Couldn't check for updates · last checked {ago}", {
        ago: FormatTime.agoText(lang, checkedAt, now)
    });
}

function statusLine(lang, check, appVersion, now) {
    switch (check.status) {
    case "up_to_date":
        return check.checkedAt === null ? I18n.tr(lang, "You're up to date · Headroom {version}", {
            version: appVersion
        }) : upToDateLine(lang, appVersion, check.checkedAt, now);
    case "available":
        return availableLine(lang, check.checkedAt, now);
    case "failed":
        return failedLine(lang, check.checkedAt, now);
    case "rate_limited":
        return rateLimitedLine(lang, check.until, now);
    case "disabled":
        return I18n.tr(lang, "Update checks are off");
    default:
        return I18n.tr(lang, "Headroom {version} · not checked yet", {
            version: appVersion
        });
    }
}

File: crates/headroom-tray/src/update_check.rs
use jiff::Timestamp;
use serde::Deserialize;

use crate::dates::{Locale, exact_moment};
use crate::format::ago_text;
use crate::i18n::fill;
use crate::payload::UpdateCheck;

const FIRST_CHECKING_DAEMON: [u64; 3] = [0, 6, 0];

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CheckStatus {
    UpToDate,
    Available,
    Failed,
    RateLimited,
    Disabled,
    #[serde(other)]
    Unknown,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct CheckOutcome {
    pub status: CheckStatus,
    #[serde(default)]
    pub checked_at: Option<Timestamp>,
    #[serde(default)]
    pub version: Option<String>,
    #[serde(default)]
    pub until: Option<Timestamp>,
}

impl CheckOutcome {
    #[must_use]
    pub fn failed(checked_at: Option<Timestamp>) -> Self {
        Self {
            status: CheckStatus::Failed,
            checked_at,
            version: None,
            until: None,
        }
    }
}

pub fn parse_outcome(json: &str) -> Result<CheckOutcome, serde_json::Error> {
    serde_json::from_str(json)
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum CheckRun {
    #[default]
    Idle,
    Checking,
    Done(CheckOutcome),
}

fn version_parts(version: &str) -> Option<[u64; 3]> {
    let core = version.trim().trim_start_matches('v');
    let core = core.split(['-', '+']).next()?;
    let mut numbers = core.split('.').map(|part| part.parse::<u64>().ok());
    Some([numbers.next()??, numbers.next()??, numbers.next()??])
}

#[must_use]
pub fn daemon_checks_on_demand(app_version: Option<&str>) -> bool {
    app_version
        .and_then(version_parts)
        .is_some_and(|parts| parts >= FIRST_CHECKING_DAEMON)
}

#[must_use]
pub fn check_row_visible(
    app_version: Option<&str>,
    update_check: Option<&UpdateCheck>,
    checks_on: bool,
) -> bool {
    checks_on && update_check.is_some() && daemon_checks_on_demand(app_version)
}

fn current_outcome(run: &CheckRun, checked_at: Option<Timestamp>) -> Option<&CheckOutcome> {
    match run {
        CheckRun::Done(outcome) if outcome.checked_at >= checked_at => Some(outcome),
        CheckRun::Idle | CheckRun::Checking | CheckRun::Done(_) => None,
    }
}

pub struct CheckContext<'a> {
    pub locale: &'a Locale,
    pub version: &'a str,
    pub checked_at: Option<Timestamp>,
    pub now: Timestamp,
}

fn checked(ctx: &CheckContext, template: &'static str, version: &str) -> String {
    let lang = ctx.locale.lang;
    match ctx.checked_at {
        Some(at) => fill(
            lang.tr(template),
            &[("version", version), ("ago", &ago_text(lang, at, ctx.now))],
        ),
        None => fill(
            lang.tr("Headroom {version} · not checked yet"),
            &[("version", version)],
        ),
    }
}

fn failed_line(ctx: &CheckContext) -> String {
    let lang = ctx.locale.lang;
    ctx.checked_at.map_or_else(
        || lang.tr("Couldn't check for updates").to_owned(),
        |at| {
            fill(
                lang.tr("Couldn't check for updates · last checked {ago}"),
                &[("ago", &ago_text(lang, at, ctx.now))],
            )
        },
    )
}

fn rate_limited_line(ctx: &CheckContext, until: Option<Timestamp>) -> String {
    let lang = ctx.locale.lang;
    until.filter(|until| *until > ctx.now).map_or_else(
        || {
            lang.tr("GitHub is limiting checks. Try again later.")
                .to_owned()
        },
        |until| {
            fill(
                lang.tr("GitHub is limiting checks until {moment}"),
                &[("moment", &exact_moment(until, ctx.now, ctx.locale))],
            )
        },
    )
}

const UP_TO_DATE: &str = "You're up to date · Headroom {version} · checked {ago}";

#[must_use]
pub fn check_line(ctx: &CheckContext, run: &CheckRun) -> String {
    if *run == CheckRun::Checking {
        return ctx.locale.lang.tr("Checking for updates…").to_owned();
    }
    let Some(outcome) = current_outcome(run, ctx.checked_at) else {
        return checked(ctx, UP_TO_DATE, ctx.version);
    };
    match outcome.status {
        CheckStatus::Available => {
            let version = outcome.version.as_deref().unwrap_or(ctx.version);
            checked(
                ctx,
                "Headroom {version} is available · checked {ago}",
                version,
            )
        }
        CheckStatus::Failed => failed_line(ctx),
        CheckStatus::RateLimited => rate_limited_line(ctx, outcome.until),
        CheckStatus::Disabled => ctx.locale.lang.tr("Update checks are off").to_owned(),
        CheckStatus::UpToDate | CheckStatus::Unknown => checked(ctx, UP_TO_DATE, ctx.version),
    }
}

#[cfg(test)]
#[path = "update_check_tests.rs"]
mod tests;
